server: add health checks and distress gossip

tbg opened the pull request:
While generating NodeStatus summaries, check for metrics that indicate a
severe store, node, or cluster-level problem. Metrics can either be
counters or gauges, and for the former we have to keep state so that we
can notify only when the counter increments (not when it's nonzero).

Flagged metrics are gossiped under a newly introduced key. These infos
are in turn picked up by the newly introduced internal `gossip_alerts`
table.

In effect, operators can monitor the `crdb_internal.gossip_alerts`
table on any node (though they'll want to do it on all nodes if there
are network connectivity issues). Similarly, it'll be straightforward
to plumb these warnings into the UI, though to the best of my knowledge
the UI can't just query `crdb_internal`, and we may expose them in
another more suitable location (this is trivial since Gossip is
available pretty much everywhere).

For starters, we only check the metrics for underreplicated and
unavailable ranges as well as liveness errors, but there is no
limitation on how elaborate these health checks can become. In fact,
they aren't limited to sourcing solely from `NodeStatus`, and in light
of #25316 we should consider alerting when nodes can't (bidirectionally)
communicate, so that operators can easily diagnose DNS or firewall
issues.

NB: I had originally envisioned polling the `node_metrics` table because
that would have allowed us to write the health checks in SQL. I had this
code but ultimately deleted it as it seemed too roundabout and less
extensible.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/25343)
<!-- Reviewable:end -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 20 of 20 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[pkg/gossip/keys.go, line 142 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25343#-LBwwVp7912hdIOD4GYY:-LBwwVp88wea6t-uDtQp:bl1jxqz) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f9c9595df5d2d4264a7f9b2eeb6f532fbfa1fcf/pkg/gossip/keys.go#L142)):*
> ```Go
> // NodeIDFromHealthAlertKey extracts the nodeID from a key created through
> // MakeNodeHealthAlertKey.
> func NodeIDFromHealthAlertKey(key string) (roachpb.NodeID, error) {
> ```

Can we dedupe this with NodeIDFromKey?

---

*[pkg/server/node.go, line 809 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25343#-LBwzXFW5yZnXML8ZUPd:-LBwzXFW5yZnXML8ZUPe:b-19o4qe) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f9c9595df5d2d4264a7f9b2eeb6f532fbfa1fcf/pkg/server/node.go#L809)):*
> ```Go
> 			}
> 
> 			// TODO(tschottdorf): decide if we want to report these to sentry. At the very least,
> ```

I think we want to report these to the reg cluster somehow but not to sentry. Maybe increment a metric whenever we encounter a failure here. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25343#-:-LBx5b-B6NB__spEC5bV:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
TFTR! I added reporting for some counter-based metrics, PTAL. I'm worried that some of these are a little too aggressive though. I only want to alert if there's definitely something weird going on in the cluster, but I'm sure some of these metrics can fire occasionally. Maybe I should introduce severities `WARNING` and `CRITICAL` right away? Or we can comment some of the likely flakers out. Or just go with it to see how bad it really is in practice.

---

Review status: 15 of 21 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/gossip/keys.go, line 142 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25343#-LBwwVp7912hdIOD4GYY:-LBy2H8n1rWm-WzrgFZg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f9c9595df5d2d4264a7f9b2eeb6f532fbfa1fcf/pkg/gossip/keys.go#L142)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Can we dedupe this with NodeIDFromKey?
</blockquote></details>

Done.

---

*[pkg/server/node.go, line 809 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25343#-LBwzXFW5yZnXML8ZUPd:-LBy2HrE6ZVikvFgefc3:bi1k233) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f9c9595df5d2d4264a7f9b2eeb6f532fbfa1fcf/pkg/server/node.go#L809)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I think we want to report these to the reg cluster somehow but not to sentry. Maybe increment a metric whenever we encounter a failure here. 
</blockquote></details>

Updated the TODO.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25343#-:-LBy2F7j33ADSqnDF7Cj:bp7r17l)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM. Let's see how this works in practice before adding severities. To the extent we can solve this by giving sensitive alerts non-zero thresholds, I think it's better to do that than introduce multiple kinds of alerts. 

---

Reviewed 6 of 6 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25343#-:-LC0-u8t6l3qBfdi5xJv:bbz1x2v)*
<!-- Sent from Reviewable.io -->

tbg committed:
server: add health checks and distress gossip

While generating NodeStatus summaries, check for metrics that indicate a
severe store, node, or cluster-level problem. Metrics can either be
counters or gauges, and for the former we have to keep state so that we
can notify only when the counter increments (not when it's nonzero).

Flagged metrics are gossiped under a newly introduced key. These infos
are in turn picked up by the newly introduced internal `gossip_alerts`
table.

In effect, operators can monitor the `crdb_internal.gossip_alerts`
table on any node (though they'll want to do it on all nodes if there
are network connectivity issues). Similarly, it'll be straightforward
to plumb these warnings into the UI, though to the best of my knowledge
the UI can't just query `crdb_internal`, and we may expose them in
another more suitable location (this is trivial since Gossip is
available pretty much everywhere).

For starters, we only check the metrics for underreplicated and
unavailable ranges as well as liveness errors, but there is no
limitation on how elaborate these health checks can become. In fact,
they aren't limited to sourcing solely from `NodeStatus`, and in light
of #25316 we should consider alerting when nodes can't (bidirectionally)
communicate, so that operators can easily diagnose DNS or firewall
issues.

NB: I had originally envisioned polling the `node_metrics` table because
that would have allowed us to write the health checks in SQL. I had this
code but ultimately deleted it as it seemed too roundabout and less
extensible.

Release note: None
